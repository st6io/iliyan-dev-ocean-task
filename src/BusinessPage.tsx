import { x } from '@xstyled/styled-components';

const BusinessPage = () => <x.div bg="skyblue">Business Page</x.div>;

export default BusinessPage;
