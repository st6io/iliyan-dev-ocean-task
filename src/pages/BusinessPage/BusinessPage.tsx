import { Props, x } from '@xstyled/styled-components';

import Layout from '@components/Layout';
import Table from '@components/Table';

import Section from './Section';
import nearbyPlaces from './nearby-places.mock.json';

import businessesData from '../businesses.mock.json';

const { name, image, address, phone, email } = businessesData[0];
const { number, street, city, zip, country } = address;

const nearbyPlacesRows = nearbyPlaces.map(({ name, address }) => ({
  id: name,
  cells: [name, address],
}));

const SectionGroup = (props: Props) => <x.div display="flex" py={8} px={7} {...props} />;

const BusinessPage = () => (
  <Layout>
    <x.img src={image} alt={`${name} cover photo`} w={1} h="40vh" objectFit="cover"></x.img>

    <x.div display="flex" justifyContent="space-between" my={8} mx={12}>
      <SectionGroup justifyContent="space-around" flex={4} mr={5}>
        <Section title="address" mr={10}>
          <Section.Text>{`${number} ${street}`}</Section.Text>

          <Section.Text>{`${city}, ${country} ${zip}`}</Section.Text>
        </Section>

        <Section title="contact">
          <Section.Text>{phone}</Section.Text>

          <Section.Text>{email}</Section.Text>
        </Section>
      </SectionGroup>

      <SectionGroup bg="white" flex={5}>
        <Section title="nearby places" flex={1}>
          <Table variant="secondary" rows={nearbyPlacesRows} w={1} />
        </Section>
      </SectionGroup>
    </x.div>
  </Layout>
);

export default BusinessPage;
