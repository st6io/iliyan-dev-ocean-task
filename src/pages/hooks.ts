import { gql, useApolloClient, useQuery } from '@apollo/client';

const BUSINESS_FRAGMENT = gql`
  fragment BusinessFragment on Business {
    id
    name
    description
    phone
    image
    email

    address {
      number
      street
      zip
      city
      country
    }
  }
`;

export const BUSINESSES_QUERY = gql`
  query Businesses {
    businesses @rest(type: "Business", path: "b/6231abada703bb67492d2b8f") {
      ...BusinessFragment
    }
  }
  ${BUSINESS_FRAGMENT}
`;

export const useBusinessesQuery = () => useQuery<{ businesses: Business[] }>(BUSINESSES_QUERY);

export const useBusinessQuery = (id?: string) => {
  // ensure businesses are fetched if not present in the cache
  const { loading, error } = useBusinessesQuery();

  const business: Business | null = useApolloClient().readFragment({
    id: `Business:${id}`,
    fragment: BUSINESS_FRAGMENT,
  });

  return {
    loading,
    error,
    data: { business },
  };
};
