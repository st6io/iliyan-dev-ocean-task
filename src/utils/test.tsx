import { ReactElement, ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { render as nativeRender } from '@testing-library/react';

const render = (ui: ReactElement, { path = '/', ...options } = {}) => {
  const RouterWrapper = ({ children }: { children: ReactNode }) => (
    <MemoryRouter initialEntries={[path]}>{children}</MemoryRouter>
  );

  return nativeRender(ui, {
    wrapper: RouterWrapper,
    ...options,
  });
};

export * from '@testing-library/react';
export { render };
