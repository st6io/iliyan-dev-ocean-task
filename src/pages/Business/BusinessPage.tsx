import { useParams } from 'react-router-dom';

import { Props, x } from '@xstyled/styled-components';

import Error from '@components/Error';
import Layout from '@components/Layout';
import LoadingPlaceholder from '@components/LoadingPlaceholder';
import Section from '@components/Section';
import Table, { LoadingPlaceholder as TablePlaceholder } from '@components/Table';

import nearbyPlaces from './nearby-places.mock.json';

import { useBusinessQuery } from '../hooks';

const nearbyPlacesRows = nearbyPlaces.map(({ name, address }) => ({
  id: name,
  cells: [name, address],
}));

const SectionGroup = (props: Props) => <x.div display="flex" py={8} px={7} {...props} />;

const BusinessPage = () => {
  const { businessId } = useParams();
  const { loading, error, data } = useBusinessQuery(businessId);

  const { name, image, address, phone, email } = data?.business || {};
  const { number, street, city, zip, country } = address || {};

  if (error) {
    return (
      <Layout>
        <Error />
      </Layout>
    );
  }

  return (
    <Layout>
      <x.div h="40vh">
        {loading ? (
          <LoadingPlaceholder />
        ) : (
          <x.img src={image} alt={`${name} cover photo`} w={1} h="100%" objectFit="cover"></x.img>
        )}
      </x.div>

      <x.div display="flex" justifyContent="space-between" my={8} mx={12}>
        <SectionGroup justifyContent="space-around" flex={4} mr={5}>
          {loading ? (
            <LoadingPlaceholder height="50%" />
          ) : (
            <>
              <Section title="address" mr={10}>
                <Section.Text>{`${number} ${street}`}</Section.Text>

                <Section.Text>{`${city}, ${country} ${zip}`}</Section.Text>
              </Section>

              <Section title="contact">
                <Section.Text>{phone}</Section.Text>

                <Section.Text>{email}</Section.Text>
              </Section>
            </>
          )}
        </SectionGroup>

        <SectionGroup bg="white" flex={5} flexDirection="column">
          {loading ? (
            <TablePlaceholder rowsCount={10} />
          ) : (
            <Section title="nearby places" flex={1}>
              <Table variant="secondary" rows={nearbyPlacesRows} w={1} />
            </Section>
          )}
        </SectionGroup>
      </x.div>
    </Layout>
  );
};

export default BusinessPage;
