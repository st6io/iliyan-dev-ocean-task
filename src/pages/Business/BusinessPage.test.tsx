import { Route, Routes } from 'react-router-dom';

import { DocumentNode, InMemoryCache } from '@apollo/client';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';

import { fireEvent, render, screen, waitFor } from '@src/utils/test';

import BusinessPage from './BusinessPage';

import mockBusinesses from '../businesses.mock.json';
import { BUSINESSES_QUERY, BUSINESS_FRAGMENT as MockBusinessFragment } from '../hooks';

jest.mock('@apollo/client', () => ({
  ...jest.requireActual('@apollo/client'),
  useApolloClient: () => ({
    readFragment: ({ id, fragment }: { id?: string; fragment: DocumentNode }) => {
      if (fragment.toString() === MockBusinessFragment.toString()) {
        const entityId = id?.replace('Business:', '');
        return mockBusinesses.find((business) => business.id === entityId);
      }
    },
  }),
}));

jest.mock('react-placeholder-loading', () => (props: any) => (
  <div>
    <span>Placeholder:</span>
    {JSON.stringify(props, null, 2)}
  </div>
));

const nearbyPlacesRows = [
  {
    id: 'Kimia',
    cells: ['Kimia', '6 Tennyson, Manchester LS9'],
  },
  {
    id: 'Pixoboo',
    cells: ['Pixoboo', '7730 Sage, Manchester IV1'],
  },
  {
    id: 'Dabvine',
    cells: ['Dabvine', '34 Mcbride, Manchester OX7'],
  },
];

const successBusinessesQueryMock: MockedResponse = {
  request: {
    query: BUSINESSES_QUERY,
  },
  result: {
    data: {
      businesses: mockBusinesses,
    },
  },
};

const renderPage = (mock = successBusinessesQueryMock) =>
  render(
    <MockedProvider mocks={[mock]} cache={new InMemoryCache()}>
      <Routes>
        <Route path="businesses/:businessId" element={<BusinessPage />} />
      </Routes>
    </MockedProvider>,
    {
      path: `/businesses/${mockBusinesses[0].id}`,
    },
  );

const waitToLoadData = () => waitFor(() => screen.findByText('address'));

describe('BusinessPage', () => {
  it('should match snapshot', async () => {
    const { container } = renderPage();

    await waitToLoadData();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render image cover with correct styles', async () => {
    renderPage();

    await waitToLoadData();

    const [, coverImg] = screen.queryAllByRole('img');

    expect(coverImg).toBeInTheDocument();
    expect(coverImg).toHaveAttribute(
      'src',
      'https://images.unsplash.com/photo-1527015175922-36a306cf0e20?crop=entropy&cs=tinysrgb&fit=crop&fm=jpg&h=600&w=1000',
    );
    expect(coverImg).toHaveAttribute('alt', 'Zazio cover photo');
    expect(coverImg).toHaveStyle({
      width: '100%',
      height: '100%',
      'object-fit': 'cover',
    });
  });

  it('should apply correct spacing between section groups', async () => {
    renderPage();

    await waitToLoadData();

    const [, img] = screen.queryAllByRole('img');
    const contentElement = img.parentNode?.nextSibling;

    expect(contentElement).toHaveStyle({
      display: 'flex',
      'justify-content': 'space-between',
    });
  });

  describe('Address', () => {
    it('should apply correct styles to section "Address"', async () => {
      renderPage();

      await waitToLoadData();

      const addressTitle = screen.queryByText('address');
      expect(addressTitle).toBeInTheDocument();
      expect(addressTitle).toHaveStyle({
        'font-size': '1.5rem',
        color: 'rgb(53, 52, 73)',
        'margin-bottom': '2rem',
        'text-transform': 'capitalize',
      });

      const addressSection = addressTitle?.parentNode;
      expect(addressSection).toHaveStyle({
        'margin-right': '2.5rem',
      });
    });

    it('should render full address', async () => {
      renderPage();

      await waitToLoadData();

      expect(screen.getByText('84586 Straubel')).toBeInTheDocument();
      expect(screen.getByText('Manchester, United Kingdom SG4')).toBeInTheDocument();
    });
  });

  describe('Contact', () => {
    it('should apply correct styles to section "Contact"', async () => {
      renderPage();

      await waitToLoadData();

      const contactTitle = screen.queryByText('contact');
      expect(contactTitle).toBeInTheDocument();
      expect(contactTitle).toHaveStyle({
        'font-size': '1.5rem',
        color: 'rgb(53, 52, 73)',
        'margin-bottom': '2rem',
        'text-transform': 'capitalize',
      });
    });
  });

  describe('Nearby Places', () => {
    it('should render section group with background', async () => {
      renderPage();

      await waitToLoadData();

      const placesTitle = screen.queryByText('nearby places');
      const sectionGroup = placesTitle?.parentNode?.parentNode;
      expect(sectionGroup).toHaveStyle({
        'background-color': 'rgb(255, 255, 255)',
      });
    });

    it('should use table with "secondary" variant', async () => {
      renderPage();

      await waitToLoadData();

      const trElements = screen.queryByRole('table')?.querySelectorAll('tr') || [];
      expect(trElements.length).toBeGreaterThan(0);
      trElements.forEach((rowElement) => {
        expect(rowElement).toHaveStyle({
          'background-color': 'rgb(248, 248, 250)',
        });
      });

      nearbyPlacesRows
        .flatMap(({ cells }) => cells)
        .forEach((cell) => {
          const cellElement = screen.queryByText(cell);
          expect(cellElement).toBeInTheDocument();
          expect(cellElement).toHaveStyle({
            color: 'rgb(146, 146, 157)',
          });
        });
    });

    it('should navigate to nearby business page on Nearby Places row click', async () => {
      renderPage();

      await waitToLoadData();

      fireEvent.click(screen.getByText('Kimia'));

      // Kimia Contact section
      expect(screen.getByText('762-896-8094')).toBeInTheDocument();
      expect(screen.getByText('bfurmonger2@ask.com')).toBeInTheDocument();
    });
  });

  it('should render loading screen', () => {
    const { container } = renderPage();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render error screen', async () => {
    const errorQueryMock = {
      request: {
        query: BUSINESSES_QUERY,
      },
      result: {
        data: undefined,
      },
      error: new Error('Something went wrong'),
    };
    renderPage(errorQueryMock);

    await screen.findByText('Something went wrong.');
    await screen.findByText('Please try again later.');
  });
});
