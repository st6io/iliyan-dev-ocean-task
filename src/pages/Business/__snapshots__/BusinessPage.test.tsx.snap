// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessPage should match snapshot 1`] = `
<div
  class="sc-cxabCf FKOKy"
>
  <header
    class="sc-efBctP eZxVPd"
  >
    <a
      href="/"
    >
      <img
        alt="home logo"
        class="sc-BeQoi"
        src="logo.png"
      />
    </a>
  </header>
  <div
    class="sc-cxabCf bsgFzP"
  >
    <div
      class="sc-cxabCf kaJeMk"
    >
      <img
        alt="Zazio cover photo"
        class="sc-BeQoi dRxsHj"
        src="https://images.unsplash.com/photo-1527015175922-36a306cf0e20?crop=entropy&cs=tinysrgb&fit=crop&fm=jpg&h=600&w=1000"
      />
    </div>
    <div
      class="sc-cxabCf cTSDup"
    >
      <div
        class="sc-cxabCf fBBkYE"
      >
        <section
          class="sc-bWXABl kZOqy"
        >
          <div
            class="sc-cxabCf sc-iJkHyd  bdVRyE"
          >
            address
          </div>
          <div
            class="sc-cxabCf sc-grVGCS  LTzaA"
          >
            84586 Straubel
          </div>
          <div
            class="sc-cxabCf sc-grVGCS  LTzaA"
          >
            Manchester, United Kingdom SG4
          </div>
        </section>
        <section
          class="sc-bWXABl"
        >
          <div
            class="sc-cxabCf sc-iJkHyd  bdVRyE"
          >
            contact
          </div>
          <div
            class="sc-cxabCf sc-grVGCS  LTzaA"
          >
            609-306-3610
          </div>
          <div
            class="sc-cxabCf sc-grVGCS  LTzaA"
          >
            igowanlock0@networkadvertising.org
          </div>
        </section>
      </div>
      <div
        class="sc-cxabCf cdUYaI"
      >
        <section
          class="sc-bWXABl kgzIGE"
        >
          <div
            class="sc-cxabCf sc-iJkHyd  bdVRyE"
          >
            nearby places
          </div>
          <table
            class="sc-WCkqM sc-kngDgl GQgxF bjOxoH"
          >
            <tbody
              class="sc-ikjQzJ"
            >
              <tr
                class="sc-jGprRt bYECWk"
              >
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  Kimia
                </td>
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  6 Tennyson, Manchester LS9
                </td>
              </tr>
              <tr
                class="sc-jGprRt bYECWk"
              >
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  Pixoboo
                </td>
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  7730 Sage, Manchester IV1
                </td>
              </tr>
              <tr
                class="sc-jGprRt bYECWk"
              >
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  Dabvine
                </td>
                <td
                  class="sc-tsFYE sc-afnQL fMNVnW RCrBC"
                >
                  34 Mcbride, Manchester OX7
                </td>
              </tr>
            </tbody>
          </table>
        </section>
      </div>
    </div>
  </div>
</div>
`;

exports[`BusinessPage should render loading screen 1`] = `
<div
  class="sc-cxabCf FKOKy"
>
  <header
    class="sc-efBctP eZxVPd"
  >
    <a
      href="/"
    >
      <img
        alt="home logo"
        class="sc-BeQoi"
        src="logo.png"
      />
    </a>
  </header>
  <div
    class="sc-cxabCf bsgFzP"
  >
    <div
      class="sc-cxabCf kaJeMk"
    >
      <div>
        <span>
          Placeholder:
        </span>
        {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": "100%"
}
      </div>
    </div>
    <div
      class="sc-cxabCf cTSDup"
    >
      <div
        class="sc-cxabCf fBBkYE"
      >
        <div>
          <span>
            Placeholder:
          </span>
          {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": "50%"
}
        </div>
      </div>
      <div
        class="sc-cxabCf cdUYaI"
      >
        <div
          class="sc-cxabCf eWXKur"
        >
          <div>
            <span>
              Placeholder:
            </span>
            {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": 60
}
          </div>
        </div>
        <div
          class="sc-cxabCf eWXKur"
        >
          <div>
            <span>
              Placeholder:
            </span>
            {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": 60
}
          </div>
        </div>
        <div
          class="sc-cxabCf eWXKur"
        >
          <div>
            <span>
              Placeholder:
            </span>
            {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": 60
}
          </div>
        </div>
        <div
          class="sc-cxabCf eWXKur"
        >
          <div>
            <span>
              Placeholder:
            </span>
            {
  "shape": "rect",
  "colorStart": "white",
  "width": "100%",
  "height": 60
}
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
